Fix error handling in "COPY <table> TO <file>".

hlinnaka opened the pull request:
If an error occurred in the segments, in a `COPY <table> TO <file>`
command, the COPY was stopped, but the error was not reported to the user.
That gave the false impression that it finished successfully, but what you
actually got was an incomplete file.

A test case is included. It uses a little helper output function that
sometimes throws an error. Output functions are fairly unlikely to fail,
but it could happen e.g. because of an out of memory error, or a disk
failure. The "COPY (SELECT ...) TO <file>" variant did not suffer from
this (otherwise, a query that throws an error would've been a much simpler
way to test this.)

The reason for this was that the code in cdbCopyGetData() that called
PQgetResult(), and extracted the error message from the result, didn't
indicate to the caller in any way that the error happened. To fix, delay
the call to PQgetResult(), to a later call to cdbCopyEnd(). cdbCopyEnd()
already had the logic to extract the error information from the PGresult,
and throw it to the user. While we're at it, refactor cdbCopyEnd a
little bit, to give the callers a nicer function signature.

I also changed a few places that used 32-bit int to store rejected row
counts, to use int64 instead. There was a FIXME comment about that. I
didn't fix all the places that do that, though, so I moved the FIXME to
one of the remaining places.

Apply to master branch only. GPDB 5 didn't handle this too well, either;
with the included test case, you got an error like this:

postgres=# copy broken_type_test to '/tmp/x';
ERROR:  missing error text

That's not very nice, but at least you get an error, even if it's not a very
good one. The code looks quite different in 5X_STABLE, so I'm not going to
attempt improving that.

Eulerizeit commented:
@adam8157 you might want to also check to see if this effects 5.x and 4.3

adam8157 commented on src/backend/cdb/cdbcopy.c:
```diff
@@ -646,21 +644,25 @@ cdbCopyEndAndFetchRejectNum(CdbCopy *c, int64 *total_rows_completed, char *abort
+	 * In COPY FROM, send the EOF to the segments. In COPY TO, we had already
+	 * received -1 from PQgetCopyData(), indicating EOF.
+	 */
+	if (!c->copy_in)
```
`COPY FROM` should be `if (c->copy_in)`, right?

hlinnaka replied:
Yes, well spotted!

It's a bit scary that everything seemed to work fine even when that was reversed. That perhaps warrants some further investigation, but I think that can be done as a separate issue/PR.

adam8157 replied:
I read the workflow and tested again, PQgetCopyData() gets`-1` from PQputCopyEnd() here exactly... We should send the EOF for `COPY TO`.

hlinnaka commented:
I refactored this some more, cleaning up the error handling further. I added a comment to top of cdbcopy.c, to explain how this should work now. Please take another look

ashwinstar commented:
@hlinnaka what's pending here? We should get this in.

hlinnaka commented:
The PR pipeline finally passed with this. I guess it's ready to go. Pushed. Thanks for the review, @adam8157!
